IPv4 multicast fixes

jukkar opened the pull request.

jukkar committed:
net: ip: Check IPv4 multicast address using big-endian

The IPv4 address in struct in_addr is in big endian so check
the multicast address value correctly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Add IPv4 multicast add, remove and lookup functions

There was no handler functions for adding, removing and looking up
IPv4 multicast addresses in the network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: context: Bind did not check IPv4 multicast address

If the user tried to bind to IPv4 multicast address, then the
operation failed and returned error.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -831,16 +831,27 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
+			}
+
+			ptr = &maddr->address.in_addr;
+
```
remove this empty line

jukkar replied:
Empty line would look actually better here, it actually is more crowded after removal.

tbursztyka commented on subsys/net/ip/net_if.c:
```diff
@@ -1623,6 +1623,95 @@ bool net_if_ipv4_addr_rm(struct net_if *iface, struct in_addr *addr)
+
+		for (i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
+			if (!iface->ipv4.mcast[i].is_used ||
+			    iface->ipv4.mcast[i].address.family != AF_INET) {
```
How an ipv4 address's family could not be of AF_INET?
And actually, according to net_if_ipv4_maddr_add() above, it's always set to AF_INET. Plus, net_if_ipv4_maddr_rm() keep the family set, in the long run all maddr (used or not) will be of this family.

Looks like to me looping on iface->ipv4.mcast could be factorize for all the 3 functions, something like: loop_and_find_mcast(the mcast table, a bool used) returning the relevant mcast that matches used parameter, or NULL otherwise. It could be generalized about cmp also, but would require 2 more param: the maddr to compare with, and cmp boolean to tell which result is expected. if maddr is null, it will not run any comparison.

Shorter code, less imbrications of for/if/for etc...

jukkar commented:
This is still WIP, I will create unit test for these in next version.

tbursztyka commented on subsys/net/ip/net_if.c:
```diff
@@ -1623,6 +1623,100 @@ bool net_if_ipv4_addr_rm(struct net_if *iface, struct in_addr *addr)
+	int i;
+
+	for (i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
+		if (need_free && iface->ipv4.mcast[i].is_used) {
```
I would have made s/need_free/is_used and made one test only here, as it was semantically better because maddr_rm or maddr_lookup for instance are looking for used ones, it's not like they "need" a non-free.
The "need" keyword here bugs me a bit. Fits maddr_add, but is_used was also proper here.

But ok, as you want.

jukkar replied:
Ok, I will change the naming here.

jukkar committed:
tests: net: Add IPv4 multicast address tests

These tests test the IPv4 multicast address add, lookup and
removal functions found for network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
@tbursztyka Added also unit tests, are you ok with this now?
